Framework/DataHandling/src/LoadSampleShape.cpp

```diff
@@ -0,0 +1,222 @@
+  } else {
+    return false; // end of file
+  }
+}
+
+/* Reads vertex from STL file and returns true if vertex is found */
+bool readSTLVertex(std::ifstream &file, V3D &vertex) {
+  std::string line;
+  if (getline(file, line)) {
+    boost::trim(line);
+    std::vector<std::string> tokens;
+    boost::split(tokens, line, boost::is_any_of(" "), boost::token_compress_on);
+    if (tokens.size() == 4 && tokens[0] == "vertex") {
+      vertex.setX(boost::lexical_cast<double>(tokens[1]));
+      vertex.setY(boost::lexical_cast<double>(tokens[2]));
+      vertex.setZ(boost::lexical_cast<double>(tokens[3]));
+      return true;
+    } else {
+      throw std::runtime_error("Error on reading STL vertex");
+    }
+  }
+  return false;
+}
+
+/* Reads triangle for STL file and returns true if triangle is found */
+bool readSTLTriangle(std::ifstream &file, V3D &v1, V3D &v2, V3D &v3) {
+
+  if (readSTLLine(file, "facet") && readSTLLine(file, "outer loop")) {
+    bool ok = (readSTLVertex(file, v1) && readSTLVertex(file, v2) &&
+               readSTLVertex(file, v3));
+    if (!ok) {
+      throw std::runtime_error("Error on reading STL triangle");
+    }
+  } else {
+    return false; // End of file
+  }
+  return readSTLLine(file, "endloop") && readSTLLine(file, "endfacet");
+}
+
+// Adds vertex to list if distinct and returns index to vertex added or equal
+uint16_t addSTLVertex(V3D &vertex, std::vector<V3D> &vertices) {
+  for (uint16_t i = 0; i < vertices.size(); ++i) {
+    if (areEqualVertices(vertex, vertices[i])) {
+      return i;
+    }
+  }
+  vertices.push_back(vertex);
+  uint16_t index = static_cast<uint16_t>(vertices.size() - 1);
+  if (index != vertices.size() - 1) {
+    throw std::runtime_error("Too many vertices in solid");
+  }
+  return index;
+}
+
+std::unique_ptr<MeshObject> readSTLMeshObject(std::ifstream &file) {
+  std::vector<uint16_t> triangleIndices;
+  std::vector<V3D> vertices;
+  V3D t1, t2, t3;
+
+  while (readSTLTriangle(file, t1, t2, t3)) {
+    // Add triangle if all 3 vertices are distinct
+    if (!areEqualVertices(t1, t2) && !areEqualVertices(t1, t3) &&
+        !areEqualVertices(t2, t3)) {
+      triangleIndices.push_back(addSTLVertex(t1, vertices));
+      triangleIndices.push_back(addSTLVertex(t2, vertices));
+      triangleIndices.push_back(addSTLVertex(t3, vertices));
+    }
+  }
+  // Use efficient constructor of MeshObject
+  std::unique_ptr<MeshObject> retVal = std::unique_ptr<MeshObject>(
+      new MeshObject(std::move(triangleIndices), std::move(vertices),
+                     Mantid::Kernel::Material()));
+  return retVal;
+}
+
+std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
+                                                   std::string &name) {
+  // Read Solid name
+  // We expect line after trimming to be "solid "+name.
+  std::string line;
+  if (getline(file, line)) {
+    boost::trim(line);
+    if (line.size() < 5 || line.substr(0, 5) != "solid") {
```

Comment:
should this be a case insensitive comparison?

Reply:
I checked the standard at http://www.fabbers.com/tech/STL_Format and found out that keywords do need to be in lower case.